django/db/models/expressions.py

```diff
@@ -253,18 +253,12 @@ def _resolve_output_field(self):
         this check. If all sources are `None`, then an error will be thrown
         higher up the stack in the `output_field` property.
         """
-        if self._output_field is None:
-            sources = self.get_source_fields()
-            num_sources = len(sources)
-            if num_sources == 0:
-                self._output_field = None
-            else:
-                for source in sources:
-                    if self._output_field is None:
-                        self._output_field = source
-                    if source is not None and not isinstance(self._output_field, source.__class__):
-                        raise FieldError(
-                            "Expression contains mixed types. You must set output_field")
+        sources_iter = filter(None.__ne__, self.get_source_fields())
```

Comment:
I'd also favor Simon's version. The current version is too cryptic, in my opinion. Also: `sources_iter` -> sources`?